crates/sui-framework/published_api.txt

```diff
@@ -0,0 +1,1360 @@
+0x3::genesis:
+	GenesisValidatorMetadata: struct
+	GenesisChainParameters: struct
+	TokenDistributionSchedule: struct
+	TokenAllocation: struct
+	create: private fun
+	allocate_tokens: private fun
+	activate_validators: private fun
+0x3::stake_subsidy:
+	StakeSubsidy: struct
+	create: package fun
+	advance_epoch: package fun
+	current_epoch_subsidy_amount: public fun
```

Comment:
This is a pretty small nit but it would be easier to use the output for this if:

- It matched Move syntax
- It wasn't hierarchical (so that when you see a diff on this file, you have all the information to hand).

```suggestion
public struct 0x3::genesis::GenesisValidatorMetadata
public struct 0x3::genesis::GenesisChainParameters
public struct 0x3::genesis::TokenDistributionSchedule
public struct 0x3::genesis::TokenAllocation
fun 0x3::genesis::create
fun 0x3::genesis::allocate_tokens
fun	0x3::genesis::activate_validators
public struct 0x3::stake_subsidy::StakeSubsidy
public(package) fun 0x3::stake_subsidy::create
public(package) fun 0x3::stake_subsidy::advance_epoch
public fun 0x3::stake_subsidy::current_epoch_subsidy_amount
```

Open question on whether it's easier to read in the file order or in alphabetical order.

Reply:
funny, because we started with something like this and then I changed my mind to the thing we see above. 

My goals/assumptions with the above format is that I want to quickly check names. Primarily I'm looking for anything with the word "test" in it :P

My worry with doing something more canonical like `public fun 0x3::stake_subsidy::current_epoch_subsidy_amount` is that my eyes have to jump to te end to parse the bit I care about first, and then jump back to the beginning to read what it is and its visibility. But maybe I'm over indexing on just how I would view it

Reply:
IMO I prefer the function name first, since that is easier to match-up against (as Todd said so you don't have to jump to the end), and really this is primarily meant for "PR-time audits" to make sure something isn't accidentally being included that shouldn't be.

Reply:
I don't mind the ordering aspect, (identifier first etc), but 

- I do think it's important to avoid a nested/indented format, because you will typically be viewing this file in a diff, and if the contents are nested, then the diff will not include the context of which module the function is in because git's ability to capture the parent nesting in the diff context is not great (and we can completely forget about it if there's multiple levels of nesting).
- Regardless of order, I would still use the same keywords (e.g. still `public(package)`, to avoid inventing new syntax for existing concepts).

Maybe you could have some kind of fixed alignment to make it easier to scan?

```
GenesisValidatorMetadata                public struct       0x3::genesis
GenesisChainParameters                  public struct       0x3::genesis
TokenDistributionSchedule               public struct       0x3::genesis
TokenAllocation                         public struct       0x3::genesis
create                                  fun                 0x3::genesis
allocate_tokens                         fun                 0x3::genesis
active_validators                       fun                 0x3::genesis
StakeSubsidy                            public struct       0x3::stake_subsidy
create                                  public(package) fun 0x3::stake_subsidy
advance_epoch                           public(package) fun 0x3::stake_subsidy
current_epoch_subsidy_amount            public fun          0x3::stake_subsidy
```

(the spacing would need to be fixed, not computed dynamically based on the size of the widest element, because otherwise you could end up with huge diffs -- I used a spacing of 40 for the name and 10 for the keywords here).

Reply:
> I do think it's important to avoid a nested/indented format, because you will typically be viewing this file in a diff, and if the contents are nested, then the diff will not include the context of which module the function is in because git's ability to capture the parent nesting in the diff context is not great (and we can completely forget about it if there's multiple levels of nesting).

Great point! Sold that we don't want the indented formatting the way we have it here.

> Regardless of order, I would still use the same keywords (e.g. still public(package), to avoid inventing new syntax for existing concepts).

Agreed. Will make this so :) 

In terms of final formatting, I'm worried about the fixed alignment due to possibly long names and the like (and causing nasty formatting). What if we did something like the following?

```
GenesisValidatorMetadata                
    public struct   
    0x3::genesis
GenesisChainParameters              
    public struct       
    0x3::genesis
TokenDistributionSchedule               
    public struct   
    0x3::genesis
TokenAllocation                         
    public struct       
    0x3::genesis
create                                  
    fun                 
    0x3::genesis
allocate_tokens                         
    fun                 
    0x3::genesis
active_validators                       
    fun                 
    0x3::genesis
StakeSubsidy                            
    public struct       
    0x3::stake_subsidy
create                                  
    public(package) fun 
    0x3::stake_subsidy
advance_epoch                           
    public(package) fun 
    0x3::stake_subsidy
current_epoch_subsidy_amount            
    public fun          
    0x3::stake_subsidy
```

This doesn't require fixed spacing, and the module it is in (while indented) will appear in any diff/diff context.

Reply:
Yep, no major concerns from me! I do find having everything on one line easier to scan and was not overly concerned about identifiers being longer than 40 characters or so (those identifies just wouldn't have any padding, and would push later fields out of line, but that should be so incredibly rare), but that's aesthetics at this point.